### PR TITLE
Update bookkeeper to 4.16

### PR DESCRIPTION
**Pull request by dmercuriali**
Updated bk dependency to 4.16
Use the new EmbeddedServer feature for the embedded bk server

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

### CODE REVIEW

File: herddb-core/src/main/java/herddb/cluster/EmbeddedBookie.java [hunk 1]
```diff
@@ -188,16 +190,17 @@ private void dumpBookieConfiguration(Path bookie_dir, org.apache.bookkeeper.conf
...
+        if (embeddedServer != null) {
             LOG.info("Apache Bookkeeper stopping");
             try {
-                bookieServer.shutdown();
-
-                bookieServer.join();
+                embeddedServer.getLifecycleComponentStack().close();
+                if (waitForBookieServiceState(Lifecycle.State.CLOSED)) {
+                    LOG.info("Apache Bookkeeper stopped");
+                }
```
Comment (eolivelli) [inline herddb-core/src/main/java/herddb/cluster/EmbeddedBookie.java:199]: what about adding a log line in case of timeout ?

Comment (dmercuriali) [reply herddb-core/src/main/java/herddb/cluster/EmbeddedBookie.java:199]: added a warning here and in the start() method

**Review by eolivelli — APPROVED** [review]

File: herddb-core/pom.xml [hunk 2]
```diff
@@ -126,7 +126,6 @@
         <dependency>
             <groupId>commons-cli</groupId>
             <artifactId>commons-cli</artifactId>
-            <scope>test</scope>
```
Comment (diegosalvi) [inline herddb-core/pom.xml:129]: Why import commons-cli not only in tests?

**Review by diegosalvi — APPROVED** [review]

File: herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java [hunk 3]
```diff
@@ -246,8 +249,8 @@ public void close() throws LogNotAvailableException {
             try {
                 LOGGER.log(Level.INFO, "{0} closing ledger {1}, with LastAddConfirmed={2}, LastAddPushed={3} length={4}, errorOccurred:{5}",
                         new Object[]{tableSpaceDescription(), out.getId(), out.getLastAddConfirmed(), out.getLastAddPushed(), out.getLength(), errorOccurredDuringWrite});
-                out.close();
-            } catch (InterruptedException | org.apache.bookkeeper.client.api.BKException err) {
+                out.closeAsync().get(60, TimeUnit.SECONDS);
+            } catch (InterruptedException | ExecutionException | TimeoutException err) {
```
Comment (eolivelli) [inline herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java:253]: usually it i s better to unwrap ExecutionException.getCause() in order to get rid of the intermediate ExecutionException.

We should also make the timeout configurable

**Review by eolivelli — COMMENTED** [review]
Nice work !

I have added some last minute feedback

@diegosalvi @dmercuriali

**Comment by eolivelli** [comment]
Also there is an ongoing VOTE for BK 4.16.2, I suggest to wait for that release before cutting the HerdDB release

cc @aluccaroni

**Comment by dmercuriali** [comment]
follow up PR #810

**Pull request merged**